modules/recursion/arrow-recursion-data/src/main/kotlin/arrow/recursion/typeclasses/Recursive.kt

```diff
@@ -1,31 +1,89 @@
 package arrow.recursion.typeclasses
 
 import arrow.Kind
-import arrow.core.Eval
+import arrow.core.*
+import arrow.core.extensions.tuple2.comonad.comonad
+import arrow.free.Cofree
+import arrow.free.CofreePartialOf
+import arrow.free.extensions.cofree.comonad.comonad
+import arrow.free.fix
+import arrow.recursion.*
+import arrow.typeclasses.Comonad
 import arrow.typeclasses.Functor
-import arrow.recursion.Algebra
-import arrow.recursion.Coalgebra
-import arrow.recursion.hylo
 
 /**
  * ank_macro_hierarchy(arrow.recursion.typeclasses.Recursive)
  *
  * Typeclass for types that can be generically folded with algebras.
  */
-interface Recursive<T> {
+interface Recursive<T, F> {
+
+  fun FF(): Functor<F>
+
   /**
    * Implementation for project.
    */
-  fun <F> Functor<F>.projectT(tf: Kind<T, F>): Kind<F, Kind<T, F>>
+  fun T.projectT(): Kind<F, T>
```

Comment:
I'm not sure these are equivalent, right? `T` isn't restricted to be a Kind like on the original.

Reply:
This basically allows any non-recursive type to be recursive if they can provide a functor f for which they can implement projectT

Reply:
If they have a `projectT` isn't that the `just` part of `Applicative`?

Reply:
Not really, applicative just will not convert structure correctly. Take lists for example (using ListF as a base functor): 
```
ListK<A>.projectT(): Kind<ListFPartialOf<A>, T> = when {
  isEmpty() -> ListF.NilF()
  else -> ListF.ConsF(first(), drop(1).k()) 
 }
```
Applicative would not be able to do that.

Reply:
as another example where applicative can't really work: The instance for long together with a fib alg.
```
interface LongBirecursive : Birecursive<Long, ForOption> {
  override fun FF(): Functor<ForOption> = Option.functor()

  override fun Kind<ForOption, Eval<Long>>.embedT(): Eval<Long> = binding {
    fix().fold({ 0L }, { it.bind() + 1L })
  }

  override fun Long.projectT(): Kind<ForOption, Long> = when {
    this < 0 -> throw IllegalArgumentException("IntBirecursive only works on natural numbers")
    this == 0L -> none()
    else -> (this - 1).some()
  }
}

fun Long.Companion.birecursive(): Birecursive<Long, ForOption> = object : LongBirecursive {}

fun fib(n: Long): Long = Long.birecursive().run {
  n.histo {
    it.fix().fold({
      0
    }, { co ->
      co.tail.value().fix().fold({
        1
      }, {
        println("(${co.head}, ${it.head})")
        co.head + it.head
      })
    })
  }
}

fun main() {
  fib(92).also(::println)
}
```